File: swift/Examples/Basic/Basic.swift
import ArgumentParser
import HelpTree

struct Basic: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "basic",
        abstract: "A basic example CLI with nested subcommands",
        subcommands: [Project.self, Task.self]
    )

    @Flag(name: .customLong("help-tree"), help: "Print a recursive command map derived from framework metadata")
    var helpTree: Bool = false

    @Option(name: [.customShort("L"), .customLong("tree-depth")], help: "Limit --help-tree recursion depth")
    var treeDepth: Int?

    @Option(name: [.customShort("I"), .customLong("tree-ignore")], help: "Exclude subtrees/commands from --help-tree output")
    var treeIgnore: [String] = []

    @Flag(name: [.customShort("a"), .customLong("tree-all")], help: "Include hidden subcommands in --help-tree output")
    var treeAll: Bool = false

    @Option(name: .customLong("tree-output"), help: "Output format (text or json)")
    var treeOutput: String?

    @Option(name: .customLong("tree-style"), help: "Tree text styling mode (rich or plain)")
    var treeStyle: String?

    @Option(name: .customLong("tree-color"), help: "Tree color mode (auto, always, never)")
    var treeColor: String?
}

struct Project: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "project",
        abstract: "Manage projects",
        subcommands: [List.self, Create.self]
    )
}

struct Task: ParsableCommand {
    static var configuration = CommandConfiguration(
        commandName: "task",
        abstract: "Manage tasks",
        subcommands: [List.self, Done.self]
    )
}

struct List: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "list", abstract: "List all items")
}

struct Create: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "create", abstract: "Create a new project")
    @Argument(help: "Project name")
    var name: String
}

struct Done: ParsableCommand {
    static var configuration = CommandConfiguration(commandName: "done", abstract: "Mark a task as done")
    @Argument(help: "Task ID")
    var id: Int
}

@main
struct BasicEntry {
    static func main() {
        let args = Array(CommandLine.arguments.dropFirst())
        if let invocation = HelpTree.parseInvocation(args) {
            HelpTree.run(for: Basic.self, invocation: invocation)
        } else {
            Basic.main()
        }
    }
}

File: rust/crates/help-tree/examples/basic.rs
use clap::{Args, Parser, Subcommand};
use help_tree::run_with_help_tree;

#[derive(Parser)]
#[command(name = "basic")]
#[command(about = "A basic example CLI with nested subcommands")]
struct Cli {
    #[arg(long, global = true, help = "Verbose output")]
    verbose: bool,

    #[command(flatten)]
    help_tree: help_tree::HelpTreeArgs,

    #[command(subcommand)]
    command: Commands,
}

#[derive(Subcommand)]
enum Commands {
    #[command(about = "Manage projects")]
    Project(ProjectArgs),
    #[command(about = "Manage tasks")]
    Task(TaskArgs),
}

#[derive(Args)]
struct ProjectArgs {
    #[command(subcommand)]
    command: ProjectCommands,
}

#[derive(Subcommand)]
enum ProjectCommands {
    #[command(about = "List all projects")]
    List,
    #[command(about = "Create a new project")]
    Create {
        #[arg(help = "Project name")]
        name: String,
    },
}

#[derive(Args)]
struct TaskArgs {
    #[command(subcommand)]
    command: TaskCommands,
}

#[derive(Subcommand)]
enum TaskCommands {
    #[command(about = "List all tasks")]
    List,
    #[command(about = "Mark a task as done")]
    Done {
        #[arg(help = "Task ID")]
        id: u32,
    },
}

fn main() {
    run_with_help_tree::<Cli>(|| {
        let cli = Cli::parse();
        match cli.command {
            Commands::Project(_) => println!("Project command"),
            Commands::Task(_) => println!("Task command"),
        }
    });
}
